### DISCUSSION TITLE
[Transfer] How to correspondingly tranfer variables between apps?

### DISCUSSION THREAD

**Discussion by Bearunner** [idaholab/moose#26211] [Q&A Modules: General] [answered]
1

**Comment by GiudGiud** [comment]
Hello

Depends on the algorithm you want to use to select a value
I would recommend this one:
https://mooseframework.inl.gov/source/transfers/MultiAppGeneralFieldShapeEvaluationTransfer.html

Guillaume

**Reply by GiudGiud** [reply]
which direction is the transfer? from pipe to volume?

**Reply by Bearunner** [reply]
The transfer happens both from pipe to volume and from volume to pipe.

**Reply by GiudGiud** [reply, marked as answer]
volume to pipe, shape evaluation will do well
pipe to volume, I would use NearestLocation OR user object with a layered average user object

**Reply by Bearunner** [reply]
thanks a lot. I will try them and let you know the results.

**Reply by Bearunner** [reply]
Hi Guillaume. 

I get the aux variable T_well (in pipe app) from the source variable T_res by using [MultiAppShapeEvaluationTransfer]. 

How can I use the aux variable T in the function /postprocessor in the sub app?

Thanks.

**Reply by GiudGiud** [reply]
If you want to use a PP in the subapp, either:
- perform some averaging using a user object then use the MultiAppGeneralFieldUserObjectTransfer
- transfer the local value of the variable to the postprocessor with 
https://mooseframework.inl.gov/source/transfers/MultiAppVariableValueSamplePostprocessorTransfer.html

**Comment by Bearunner** [comment]


**Reply by GiudGiud** [reply]
Functions cannot use variables btw. Not even ParsedFunctions. Do you mean a material property?
The first step with making a multiapp simulation is getting both inputs to work separately.

Maybe list exactly for both transfers what you plan to transfer, and what form it is in MOOSE (auxvariable? PPs?)

**Comment by Bearunner** [comment]


**Reply by GiudGiud** [reply]
Hello

No problem with 1.

2)
```
of every volume element
These Tem_res will be stored in the postprocessor or function and finally used in the heat kernels of the sub-app.
```
This does not seem consistent. In the first you say `of every volume element`, so you likely want to project the 3D field onto the 1D pipe. But then you mention using a postprocessor, which does not let you capture multiple values. Only 1 value in a postprocessor.
Do you mean a vector postprocessor that is indexed by 1D elements? Have you tried a 1D variable instead?

Functions are also not the right medium to hold a temperature field. Functions are meant to be dependent only on X Y Z and time. 

3)
Same problem

I would consider transferring to variables instead of postpcessors. 

Guillaume

**Reply by Bearunner** [reply]
thanks for your help!

**Comment by Bearunner** [comment]


**Reply by GiudGiud** [reply]
Hello

I'm actually in charge of developing these transfers so if you can make a MWE that shows that you are not getting the expected value I can look at this. 

If the 1D domain overlaps on the 3D domain, I dont see why it does not work to evaluarte the shape function.
Do you have a hole in the 3D domain where the 1D domain is?

Guillaume

**Reply by Bearunner** [reply]
In reality, we have a hole in the 3D domain. But the radius of our 1D pipe model is rather smaller than the block size in 3D model. 

So we simply ignore the hole in 3D.

**Reply by GiudGiud** [reply]
Ok so in 3D you should be able to evaluate the variables at the location of the 1D pipe right? because it does overlap.
so why does it not work for you?

**Reply by GiudGiud** [reply]
can you try:
- greedy search for the transfer
- making sure the 1D domain fits entirely within the region (reduce its length by 0.1 for example) that provides the values (block 1?)

**Reply by Bearunner** [reply]
Yes. The 'greedy_search = true' is set as above the [./to_pro_sub_P] in [Transfers].
And The 1D pipe has a length of 90m fitting entirely within the region.

**But I still get a similar error.**


(1) 1D pipe

[Mesh]
 [./file]
  type = GeneratedMeshGenerator
  dim = 1
  xmin = -2995
  xmax = 0
  nx = 600
 [../]

 [./sink_area]
  type = SubdomainBoundingBoxGenerator
  input = file
  block_id = '1'
  bottom_left = '-2905 -0.1 -0.1'
  top_right = '-2995 0.1 0.1'
 [../]

[]


_*** ERROR ***
The following error occurred in the object "to_pro_sub_rho", of type "MultiAppGeneralFieldShapeEvaluationTransfer".

Element 17 for app 0 could not be located_

**Reply by GiudGiud** [reply]
I m not sure I understand.
3D extends in [-3000, -2900] x [95, 100] x [-2.5, 2.5]
1D extends in [-2995, -2905] x [195.1,194.9] x [-0.1,0.1] (from `positions = '0 195 0'` and the 1D generator)

It seems to me like they dont overlap at all in Y?

**Reply by Bearunner** [reply]
Thanks for your tips. Now I can use MultiAppGeneralFieldShapeEvaluationTransfer for tranfering 3D models' data to pipe.

Unfortunally, I still get the following wrong temperature results around the production well.

**Reply by GiudGiud** [reply]
so this is the 3D model results right? So the problem is coming from the 1D to 3D transfer maybe?
What does the output of this transfer look like?

**Reply by GiudGiud** [reply]
so that output looks fine then?
Looks good to me. 
What could be affecting the temperature like this?
Can you simplify the problem?

Also is your mesh well-made? That s always my concern. You can use a MeshDiagnosticsGenerator to check it

**Reply by GiudGiud** [reply]
Interesting.
@cpgr might have a solution for this. 

I dont know how we can create sidedness in either the 1D object or the 3D mesh. Maybe in the 3D mesh there could be a slit / cut, and boundary conditions on both sides

**Reply by Bearunner** [reply]
Thanks for your tips.

Dear @cpgr Chris, could you please help me on this? Thank you.

**Reply by cpgr** [reply]
Hmm, I'm not sure what the solution is for this problem

**Reply by GiudGiud** [reply]
The strong sidedness around the 1D problem, how does it appear?

How do you get the velocity to be very different on both sides? And should there be velocity across the well?
